src/charm.py

```diff
@@ -216,6 +249,48 @@ def _on_cluster_relation_changed(self, event: EventBase) -> None:  # noqa: C901
         if not self.state.unit_server.started:
             self.init_server()
 
+        # create services if we expose the charm, no op if not
+        self.update_external_services()
```

Comment:
question: If the value of `expose-external` changes only by the change in the config option, wouldn't it suffice triggering this function only once in `config-changed` handler?

Reply:
This method is the "reconcile everything if something changes" method, notably called by `config-changed` on L137. 
What we could do, though, is to add a new field to the peer relation application databag (`{"expose": <current_service>}` or something), which would be used to shunt this call if not needed. This is something I have in mind but not yet implemented because I wanted first to test the current solution as-is.

Adding this new field does not solve all the issues because we still have to update the service if we use `loadbalancer-extra-annotations`, so that means a new databag field, and I want to avoid solving my problems with the "one issue, one new databag field" solution.

Reply:
Do note that the 'current-service' might be out of date, so we'd still need to call it anyway just in case I think